### DISCUSSION TITLE
How to improve the data transfer rate?

### DISCUSSION THREAD

**Discussion by f-zl** [cesanta/mongoose#2378] [Ask questions, get answers]
I'm developing an http server on an embedded system running FreeRTOS and lwip. The server waits for a single POST request that uploads a boot image file. The file size is about 4.8MB.
The problem is the low data transfer rate, about 1KB/s. Wireshark shows that it takes more than 100 ms for the server to respond to (ACK) a TCP packet. And the time gets longer and longer as the data transfers.
If I replace the mongoose server with a custom TCP server using lwip, the transfer rate is about 800KB/s, so the problem is not in hardware or lwip. Please give me some advices about how to improve the data transfer rate, thanks.

My config:
```c
#define MG_ARCH MG_ARCH_FREERTOS
#define MG_ENABLE_LWIP 1
#define MG_ENABLE_DIRLIST 0
#define MG_ENABLE_FILE 0
#define MG_MAX_RECV_SIZE (5 * 1024 * 1024)
```

My code:
```c
static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
	(void) fn_data;
	switch (ev) {
	case MG_EV_HTTP_MSG:
		HandleHttpMsg(c, (struct mg_http_message *) ev_data);
		break;
	default:
		break;
	}
}
void ServerTask(void *p) {
	(void) p;
	struct mg_mgr mgr;
	mg_mgr_init(&mgr);
	mg_http_listen(&mgr, s_listen_on, fn, NULL);
	for (;;)
		mg_mgr_poll(&mgr, 10);
	mg_mgr_free(&mgr);
}
```

**Comment by f-zl** [comment]
I'm suspecting the cause is the frequent call of calloc. Is there a way to use statically allocated memory to store the packet?

**Comment by scaprile** [comment]
If after reading the guidelines in the [documentation](https://mongoose.ws/documentation), including [the best practices section](https://mongoose.ws/documentation/#best-practices), and our [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/), like https://mongoose.ws/documentation/tutorials/file-uploads/ in particular, you still can't solve your issue:
Please open an issue, **fill the issue template**, and describe _exactly_ what you do beyond `HandleHttpMsg(c, (struct mg_http_message *) ev_data);`; what do you do with your data ? Do you answer ? Do you free ? Are you just letting data accumulate and a buffer designed for holding a small piece of data during a short time constantly resize because you do not take data out and buffer it yourself as designed and explained in the tutorials and try to handle a whole big file instead of small chunks ?

**Reply by f-zl** [reply]
`HandleHttpMsg` is an empty function by now, and it will be a function that writes the received POST body (a boot image file) to a flash memory. It takes more than an hour before HandleHttpMsg is called so that doesn't matter. I try `#define MG_IO_SIZE (5*1024*1024)` based on cpq's advice, and it improves the data rate a lot. But that increase heap usage to 2*MG_IO_SIZE when receiving that request. I will try using HTTP chuncked encoding and accumulate the body in an static buffer in the `fn` callback.

**Comment by cpq** [comment]
@f-zl You can also increase `MG_IO_SIZE`
